sdk/iot/hub/samples/src/paho_iot_hub_c2d_example.c

```diff
@@ -264,40 +264,53 @@ int main()
     printf("Failed to create MQTT client, return code %d\n", rc);
     return rc;
   }
-
-  // Set the callback for incoming MQTT messages
-  if ((rc = MQTTClient_setCallbacks(mqtt_client, NULL, NULL, on_received, NULL))
-      != MQTTCLIENT_SUCCESS)
-  {
-    printf("Failed to set MQTT callbacks, return code %d\n", rc);
-    return rc;
-  }
-
-  // Connect to IoT Hub
-  if ((rc = connect_device()) != 0)
-  {
-    return rc;
-  }
-
-  // Subscribe to the C2D topic to receive cloud-to-device messages
-  if ((rc = subscribe()) != 0)
-  {
-    return rc;
-  }
-
-  printf("Subscribed to topics.\n");
-
-  // Wait for any incoming C2D messages
-  printf("Waiting for activity. [Press ENTER to abort]\n");
-  (void)getchar();
-
-  if ((rc = MQTTClient_disconnect(mqtt_client, TIMEOUT_MQTT_DISCONNECT_MS)) != MQTTCLIENT_SUCCESS)
+  else
   {
-    printf("Failed to disconnect MQTT client, return code %d\n", rc);
-    return rc;
+    // Set the callback for incoming MQTT messages
+    if ((rc = MQTTClient_setCallbacks(mqtt_client, NULL, NULL, on_received, NULL))
+        != MQTTCLIENT_SUCCESS)
+    {
+      //Failed to set the callback - move to destroy mqtt client and return
+      printf("Failed to set MQTT callbacks, return code %d\n", rc);
+    }
+    else
+    {
+      // Connect to IoT Hub
+      if ((rc = connect_device()) != 0)
```

Comment:
Here, always attempting to do cleanup is bad practice. If the error is indicative of memory corruption, out of memory or access violation.
Example:  MQTTClient_Connect could throw `PAHO_MEMORY_ERROR`. Handling this will:
1. likely lead to even more corruption within the cleanup phase.
2. fail due to out of memory within printf and other library functions.

In general, teaching our customers to _always attempt to recover from all errors_ may lead in almost impossible to understand issues in production settings.

Reply:
Another example: `MQTTCLIENT_BAD_STRUCTURE` is (outside of dev-time errors) indicative of memory corruption as the internal struct is missing the expected signature.